### PR TITLE
Update Subscriptions article to use graphql-ws library

### PR DESCRIPTION
**Pull request by rkoron007**
First pass at updating this article to use the `graphql-ws` library. 

Much thanks @glasser for [providing an example](https://github.com/apollographql/docs-examples/tree/main/apollo-server/v3/subscriptions-graphql-ws) to work from!

### CODE REVIEW

File: docs/source/data/subscriptions.mdx [hunk 1]
```diff
@@ -540,6 +548,17 @@ The following are community-created `PubSub` libraries for popular event-publish
...
+
+This article details how to setup the [`graphql-ws`](https://www.npmjs.com/package/graphql-ws) library to add support for subscriptions to Apollo Server. Previously this article recommended using the `subscriptions-transport-ws` library, however this library is no longer actively maintained.
+
+Apollo Server supports both `graphql-ws` and `subscriptions-transport-ws`, but we strongly recommend using the currently maintained `graphql-ws` library.
+
+Both `graphql-ws` and `subscriptions-transport-ws` implement different protocols for GraphQL subscriptions over websockets. Meaning, if you do decide to switch from one library to another you will need to ensure your [Apollo Client subscription protocol](https://www.apollographql.com/docs/react/data/subscriptions/#choice-of-subscription-protocol) is **consistent** with the subscription library you are using within Apollo Server and vice versa.
+
+Note that you might need to [adjust the Apollo Sandbox settings to use the same subscription protocol as the subscription library you are using](https://www.apollographql.com/docs/studio/explorer/additional-features/#subscription-support).
```
Comment (glasser) [inline docs/source/data/subscriptions.mdx:561]: Also I'd suggest replacing this with a brief section (formatted however makes sense, bulleted list?) explaining how to use graphql-ws with each part of the platform:
- settings for Explorer
- AC web: use (not yet released) AC 3.6 (https://github.com/apollographql/apollo-client/pull/9067/files) and GraphQLWsLink instead of WebSocketLink (https://www.apollographql.com/docs/react/api/link/apollo-link-subscriptions/ will be the doc link once released; here's a preview: https://github.com/apollographql/apollo-client/pull/9067/files)
- AC Kotlin: https://www.apollographql.com/docs/kotlin/essentials/subscriptions/#customizing-your-websocket-protocol
- AC iOS: not yet supported

File: docs/source/data/subscriptions.mdx [hunk 2]
```diff
@@ -47,139 +49,148 @@ subscription PostFeed {
...
+        path: '/graphql',
+      });
+
+      // Hand in the schema we just created and have the
+      // WebSocketServer start listening.
+      useServer({ schema }, wsServer);
     ```
 
-6. Add a plugin to your `ApolloServer` constructor to close the `SubscriptionServer`.
+6. Add [a plugin](../integrations/plugins) to your `ApolloServer` constructor to close the `WebSocketServer`:
```
Comment (glasser) [inline docs/source/data/subscriptions.mdx:106]: Oh huh, I need to make sure to find the right way to do this in my example. I don't know if this is the right API for graphql-ws.

File: docs/source/data/subscriptions.mdx [hunk 3]
```diff
@@ -366,159 +379,154 @@ subscription IncrementingNumber {
...
-
-###### `webSocket`
-
-`WebSocket`
-</td>
-<td>
-
-The connecting or disconnecting [`WebSocket`](https://developer.mozilla.org/en-US/docs/Web/API/WebSocket).
+```ts
+const getDynamicContext = (ctx, msg, args) => {
```
Comment (glasser) [inline docs/source/data/subscriptions.mdx:418]: This needs to be declared as an `async` function to use `await`

**Review by glasser — CHANGES_REQUESTED** [review]
Looking great in general!

File: docs/source/data/subscriptions.mdx (hunk 2 above)
Comment (glasser) [reply docs/source/data/subscriptions.mdx:106]: Fixed this in the example in https://github.com/apollographql/docs-examples/pull/40

Note that this involved not just draining the subscription server, but also draining the main HTTP server! Also had to reorder some stuff so that things got defined in the right order.

File: docs/source/data/subscriptions.mdx (hunk 1 above)
[the remark this answers is not in the document]
Comment (rkoron007) [reply docs/source/data/subscriptions.mdx:561]: Oh, great catch! Thank you 🙏

File: docs/source/data/subscriptions.mdx (hunk 2 above)
Comment (rkoron007) [reply docs/source/data/subscriptions.mdx:106]: Oooh thank you for updating the example! I'll update the set up instructions in this article to follow suit. :)

File: docs/source/data/subscriptions.mdx [hunk 4]
```diff
@@ -366,159 +393,154 @@ subscription IncrementingNumber {
...
-});
-```
-
-This is especially important because metadata like auth tokens are sent differently depending on the transport.
-
-## `onConnect` and `onDisconnect`
+**For subscriptions,** you can extract information from a client's request by adding [options](https://github.com/enisdenjo/graphql-ws/blob/master/docs/interfaces/server.ServerOptions.md) to the first argument passed to the `useServer` function.
 
-You can define functions that the subscription server executes whenever a subscription request connects (`onConnect`) or disconnects (`onDisconnect`).
+For example, you can supply a [`context` property](https://github.com/enisdenjo/graphql-ws/blob/master/docs/interfaces/server.ServerOptions.md#context) to add values of your GraphQL operation [`context`](./resolvers/#the-context-argument):
```
Comment (StephenBarlow) [inline docs/source/data/subscriptions.mdx:406]: supply -> provide

"add values of" -> "add values to", yeah?

File: docs/source/data/subscriptions.mdx [hunk 5]
```diff
@@ -366,159 +393,154 @@ subscription IncrementingNumber {
...
-      <th>Name /<br/>Type</th>
-      <th>Description</th>
-    </tr>
-  </thead>
+Now here is where things might get a little bit confusing, notice above that the `context` property's first argument is `ctx`, also known as context. _What?_ The [`Context`](https://github.com/enisdenjo/graphql-ws/blob/master/docs/interfaces/server.Context.md) type in `graphql-ws` is **not the same thing** as the GraphQL `context` argument passed to your resolvers.
 
-<tbody>
-<tr>
-<td>
+The `Context` type from `graphql-ws` is where information about your subscription server and its requests resides. You can access incoming parameters from a client `subscription` request by checking the `Context.connectionParams`.
```
Comment (StephenBarlow) [inline docs/source/data/subscriptions.mdx:427]: If the above suggestion looks good, this might work here:

You can access a client's WebSocket connection parameters (such as an auth token) via the `ctx.connectionParams` property.

File: docs/source/data/subscriptions.mdx [hunk 6]
```diff
@@ -366,159 +393,154 @@ subscription IncrementingNumber {
...
+     context: (ctx, msg, args) => {
+       // Returning an object will add that information to our
+       // GraphQL context, which all of our resolvers have access to.
+      return getDynamicContext(ctx, msg, args);
+    },
+  },
+  wsServer,
+);
+```
+Putting it all together, if the `useServer.context` function returns an _object_ that object is passed to the GraphQL operation `context`, which is available to your resolvers.
```
Comment (StephenBarlow) [inline docs/source/data/subscriptions.mdx:455]: Comma after the italicized _object_

File: docs/source/data/subscriptions.mdx [hunk 7]
```diff
@@ -366,159 +393,154 @@ subscription IncrementingNumber {
...
 
-<tr>
-<td>
+Note that the `context` option is called once per `subscription` request, not once per event emission. Meaning that in the above example, every time a client sends a `subscription` operation we would check their authentication token.
```
Comment (StephenBarlow) [inline docs/source/data/subscriptions.mdx:457]: "Meaning" -> "This means"

"operation we would check" -> "operation, we check"

File: docs/source/data/subscriptions.mdx [hunk 8]
```diff
@@ -366,159 +393,154 @@ subscription IncrementingNumber {
...
 
-###### `context`
+### `onConnect` and `onDisconnect`
 
-`ConnectionContext`
-</td>
-<td>
+Expanding on the above, you can add additional options for the subscription server to execute whenever a client connects (`onConnect`) or disconnects (`onDisconnect`).
```
Comment (StephenBarlow) [inline docs/source/data/subscriptions.mdx:461]: "Expanding on the above..." -> "You can configure the subscription server's behavior whenever a client..."

File: docs/source/data/subscriptions.mdx [hunk 9]
```diff
@@ -366,159 +393,154 @@ subscription IncrementingNumber {
...
 
-Context object for the WebSocket connection. This is _not_ the `context` object for the associated subscription operation.
+Defining an `onConnect` function lets you reject a particular incoming connection by returning false or by throwing an exception. This can be helpful if you'd like to check authentication when a client first connects to your subscription server.
```
Comment (StephenBarlow) [inline docs/source/data/subscriptions.mdx:463]: "lets you" -> "enables you to"

Codefont `false`

"you'd like to" -> "you want to"

File: docs/source/data/subscriptions.mdx [hunk 10]
```diff
@@ -366,159 +393,154 @@ subscription IncrementingNumber {
...
+    },
+  },
+  wsServer,
+);
+```
+For more information and examples of using `onConnect` and `onDisconnect`, see the [`graphql-ws` documentation](https://github.com/enisdenjo/graphql-ws#ws-auth-handling).
 
-</tbody>
-</table>
+### Example: Authentication with the Server and Client
```
Comment (StephenBarlow) [inline docs/source/data/subscriptions.mdx:488]: lowercase "server" and "client"

File: docs/source/data/subscriptions.mdx [hunk 11]
```diff
@@ -366,159 +393,154 @@ subscription IncrementingNumber {
...
 
-### Example: Authentication with `onConnect`
+>If you plan to use subscriptions with Apollo Client, ensure **[both your client and server subscription protocols are consistent](https://www.apollographql.com/docs/react/data/subscriptions/#choice-of-subscription-protocol)** for the subscription library you are using (this example uses the `graphql-ws` library).
```
Comment (StephenBarlow) [inline docs/source/data/subscriptions.mdx:490]: This is an important-enough consideration that it might be a good idea to have a dedicated h2 section called like "Configuring clients" that first states at a higher level that you want to make sure all of your clients use the `graphql-ws` protocol. And then has a bulleted list item for each Apollo client library that either links to relevant configuration _or_ says that no configuration is required

File: docs/source/data/subscriptions.mdx (hunk 10 above)
Comment (StephenBarlow) [reply docs/source/data/subscriptions.mdx:488]: Actually why don't we go with "Example: Authentication with Apollo Client"

File: docs/source/data/subscriptions.mdx [hunk 12]
```diff
@@ -366,159 +393,154 @@ subscription IncrementingNumber {
...
 
-On the client, `SubscriptionsClient` supports adding information to `connectionParams` ([example](https://www.apollographql.com/docs/react/data/subscriptions/#5-authenticate-over-websocket-optional)) that will be sent with the first WebSocket message. On the server, all GraphQL subscriptions are delayed until the connection has been fully authenticated and the `onConnect` callback returns a truthy value.
+On the client side, the `GraphQLWsLink` constructor supports adding information to `connectionParams` ([example](https://www.apollographql.com/docs/react/data/subscriptions/#5-authenticate-over-websocket-optional)). Those `connectionParams` are sent to your server when it connects, allowing you to extract information from the client request by accessing `Context.connectionParams`.
```
Comment (StephenBarlow) [inline docs/source/data/subscriptions.mdx:492]: "On the client side" -> "In Apollo Client"

Comment (StephenBarlow) [reply docs/source/data/subscriptions.mdx:492]: (We just want to be a little more Apollo-specific here because maybe someone's using our server but another library for client)

File: docs/source/data/subscriptions.mdx [hunk 13]
```diff
@@ -366,159 +393,154 @@ subscription IncrementingNumber {
...
   connectionParams: {
-    authorization: clientToken,
+    authorization: user.authToken,
   },
-})
+}));
 ```
 
-The `connectionParams` argument in the `SubscriptionServer`'s `onConnect` callback contains the information passed by the client and can be used to validate user credentials.
+The `connectionParams` argument, containing the information provided by the client, will be passed to your server allowing you to validate the user's credentials.
```
Comment (StephenBarlow) [inline docs/source/data/subscriptions.mdx:508]: Let's put the "containing the..." in parens instead of commas like:

The `connectionParams` argument (which contains the information provided by the client) is passed to...


"to your server allowing you" -> "to your server, enabling you"

File: docs/source/data/subscriptions.mdx [hunk 14]
```diff
@@ -366,159 +393,154 @@ subscription IncrementingNumber {
...
 
-From the `onConnect` function, we can then return an object which is passed into our resolvers as the `context` argument during execution. For our example, we can use the `authorization` token provided by the client to look up the related user and pass that user along to our resolvers.
+From there you can use the [`useServer.context` property](#operation-context) to try and authenticate the user, returning an object that will be passed into your resolvers as the `context` argument during execution.
```
Comment (StephenBarlow) [inline docs/source/data/subscriptions.mdx:510]: I think you can cut "try and"

"that will be passed" -> "that's passed"

File: docs/source/data/subscriptions.mdx [hunk 15]
```diff
@@ -366,159 +393,154 @@ subscription IncrementingNumber {
...
-  { server, path }
-});
+  wsServer,
+);
 ```
 
-The example above looks up a user based on the auth token that is sent with the first initialization message on the transport, then returns the user object as a `Promise`. The user object found will be available as `context.currentUser` in your GraphQL resolvers.
-
-In case of an authentication error, the `Promise` will be rejected, which prevents the client's connection.
+To sum up, the example above looks up a user based on the authentication token sent with each subscription request before returning the user object to be used by our resolvers. If no user exists, or if lookup fails, our resolvers can throw an error and the corresponding `subscription` operation is not executed.
```
Comment (StephenBarlow) [inline docs/source/data/subscriptions.mdx:543]: "If no user exists or the lookup otherwise fails,"

File: docs/source/data/subscriptions.mdx [hunk 16]
```diff
@@ -540,6 +562,22 @@ The following are community-created `PubSub` libraries for popular event-publish
 
 If none of these libraries fits your use case, you can also create your own `PubSubEngine` subclass. If you create a new open-source library, click **Edit on GitHub** to let us know about it!
 
-## The `graphql-ws` transport library
+## Old `subscriptions-transport-ws` library
+
+>If you use subscriptions with Apollo Client you must ensure **[both your client and server subscription protocols are consistent](https://www.apollographql.com/docs/react/data/subscriptions/#choice-of-subscription-protocol)** for the subscription library you are using.
+
+This article details how to set up the [`graphql-ws`](https://www.npmjs.com/package/graphql-ws) library to add support for subscriptions to Apollo Server. Previously this article recommended using the `subscriptions-transport-ws` library; however, this library is no longer actively maintained.
```
Comment (StephenBarlow) [inline docs/source/data/subscriptions.mdx:569]: details -> describes (I think we should only use "details" as a noun for more predictability)

Comment (StephenBarlow) [reply docs/source/data/subscriptions.mdx:569]: Actually, I think we can cut the first sentence entirely, and tweak the rest to:

This article previously demonstrated using the `subscriptions-transport-ws` library to set up subscriptions. However, this library is no longer actively maintained. You can still use it with Apollo Server, but we strongly recommend using `graphql-ws` instead.

File: docs/source/data/subscriptions.mdx [hunk 17]
```diff
@@ -540,6 +562,22 @@ The following are community-created `PubSub` libraries for popular event-publish
 
 If none of these libraries fits your use case, you can also create your own `PubSubEngine` subclass. If you create a new open-source library, click **Edit on GitHub** to let us know about it!
 
-## The `graphql-ws` transport library
+## Old `subscriptions-transport-ws` library
```
Comment (StephenBarlow) [inline docs/source/data/subscriptions.mdx:565]: I might call this section "Switching from `subscriptions-transport-ws`"

File: docs/source/data/subscriptions.mdx [hunk 18]
```diff
@@ -540,6 +562,22 @@ The following are community-created `PubSub` libraries for popular event-publish
...
+
+>If you use subscriptions with Apollo Client you must ensure **[both your client and server subscription protocols are consistent](https://www.apollographql.com/docs/react/data/subscriptions/#choice-of-subscription-protocol)** for the subscription library you are using.
+
+This article details how to set up the [`graphql-ws`](https://www.npmjs.com/package/graphql-ws) library to add support for subscriptions to Apollo Server. Previously this article recommended using the `subscriptions-transport-ws` library; however, this library is no longer actively maintained.
+
+Apollo Server can be combined with both `graphql-ws` and `subscriptions-transport-ws`, but we strongly recommend using the currently maintained `graphql-ws` library.
+
+If you intend to switch from `subscriptions-transport-ws` to `graphql-ws` you will need to:
+* [Adjust Apollo Studio Explorer settings](https://www.apollographql.com/docs/studio/explorer/additional-features/#subscription-support)
+* Migrate to [Apollo Client 3.6 or later](https://www.apollographql.com/docs/react/data/subscriptions/#choice-of-subscription-protocol)
```
Comment (StephenBarlow) [inline docs/source/data/subscriptions.mdx:575]: Migrate -> Update

File: docs/source/data/subscriptions.mdx (hunk 11 above)
Comment (StephenBarlow) [reply docs/source/data/subscriptions.mdx:490]: Ah ok so this is covered in the bottom section! Maybe just have this notice link down there since it's a pretty good summary of considerations?

File: docs/source/data/subscriptions.mdx (hunk 4 above)
Comment (rkoron007) [reply docs/source/data/subscriptions.mdx:406]: Oooh yes, good catch!

File: docs/source/data/subscriptions.mdx (hunk 10 above)
Comment (rkoron007) [reply docs/source/data/subscriptions.mdx:488]: I like that!

File: docs/source/data/subscriptions.mdx (hunk 11 above)
Comment (rkoron007) [reply docs/source/data/subscriptions.mdx:490]: Such a great idea, thank you!

File: docs/source/data/subscriptions.mdx (hunk 5 above)
Comment (rkoron007) [reply docs/source/data/subscriptions.mdx:427]: I think the phrase "WebSocket connection parameters" feels a bit wordy. How about a slight rephrase:

"You can access the parameters of a client's `subscription` request to your WebSocket server via the `ctx.connectionParams` property."

Comment (StephenBarlow) [reply docs/source/data/subscriptions.mdx:427]: Yep that looks good!

**Review by glasser — APPROVED** [review]
The stuff I was looking for looks good!

(Note that we are still blocking merge until at least the release of Apollo Client 3.6. But we can get it ready to go before then.)

File: docs/source/data/subscriptions.mdx [hunk 19]
```diff
@@ -540,6 +562,20 @@ The following are community-created `PubSub` libraries for popular event-publish
...
-## The `graphql-ws` transport library
+## Switching from `subscriptions-transport-ws`
+
+>If you use subscriptions with Apollo Client you must ensure **[both your client and server subscription protocols are consistent](https://www.apollographql.com/docs/react/data/subscriptions/#choice-of-subscription-protocol)** for the subscription library you are using.
+
+This article previously demonstrated using the `subscriptions-transport-ws` library to set up subscriptions. However, this library is no longer actively maintained. You can still use it with Apollo Server, but we strongly recommend using [`graphql-ws`](https://www.npmjs.com/package/graphql-ws) instead.
+
+If you intend to switch from `subscriptions-transport-ws` to `graphql-ws` you will need to:
+* [Adjust Apollo Studio Explorer settings](https://www.apollographql.com/docs/studio/explorer/additional-features/#subscription-support)
+* Update to [Apollo Client 3.6 or later](https://www.apollographql.com/docs/react/data/subscriptions/#choice-of-subscription-protocol)
```
Comment (glasser) [inline docs/source/data/subscriptions.mdx:573]: @rkoron007 It's actually 3.5.10, which is out of today!

(Apollo iOS should be close to done as well and then we can move forward with this PR!)

File: docs/source/data/subscriptions.mdx [hunk 20]
```diff
@@ -366,180 +393,305 @@ subscription IncrementingNumber {
...
 
-###### `connectionParams`
+[Apollo Studio Explorer](https://www.apollographql.com/docs/studio/explorer/explorer/)
 
-`Object`
+</td>
+
+<td>
 
+N/A
```
Comment (glasser) [inline docs/source/data/subscriptions.mdx:593]: maybe just blank

File: docs/source/data/subscriptions.mdx [hunk 21]
```diff
@@ -366,180 +393,305 @@ subscription IncrementingNumber {
...
 </td>
+
 </tr>
 
+
 <tr>
 <td>
 
-###### `context`
+[Apollo iOS](https://www.apollographql.com/docs/studio/explorer/additional-features/#subscription-support)
```
Comment (glasser) [inline docs/source/data/subscriptions.mdx:638]: wrong link

File: docs/source/data/subscriptions.mdx [hunk 22]
```diff
@@ -366,180 +393,305 @@ subscription IncrementingNumber {
...
-  connectionParams: {
-    authorization: clientToken,
-  },
-})
-```
+<tr>
+<td>
 
-The `connectionParams` argument in the `SubscriptionServer`'s `onConnect` callback contains the information passed by the client and can be used to validate user credentials.
+[Apollo Kotlin](https://www.apollographql.com/docs/studio/explorer/additional-features/#subscription-support)
```
Comment (glasser) [inline docs/source/data/subscriptions.mdx:664]: wrong link

File: docs/source/data/subscriptions.mdx [hunk 23]
```diff
@@ -366,180 +393,305 @@ subscription IncrementingNumber {
...
-- [Kafka](https://github.com/ancashoria/graphql-kafka-subscriptions)
-- [Postgres](https://github.com/GraphQLCollege/graphql-postgres-subscriptions)
-- [Google Cloud Firestore](https://github.com/MrBoolean/graphql-firestore-subscriptions)
-- [Ably Realtime](https://github.com/ably-labs/graphql-ably-pubsub)
-- [Google Firebase Realtime Database](https://github.com/swantzter/graphql-firebase-subscriptions)
+</tbody>
+</table>
 
-If none of these libraries fits your use case, you can also create your own `PubSubEngine` subclass. If you create a new open-source library, click **Edit on GitHub** to let us know about it!
+<!-- Just going to comment these out for now -->
```
Comment (glasser) [inline docs/source/data/subscriptions.mdx:689]: Remove this now

**Review by glasser — APPROVED** [review]
With the changes below, plus rebasing/merging due to a recent change to the file, I think you could merge!

**Review by glasser — APPROVED** [review]
